test/config/config.test.js

```diff
@@ -596,6 +600,267 @@ const testFixtures = [
+        err,
+        `pkg-zero-conf-nsname/index.js script succeeded: err=${err}`,
+      );
+      const lines = stdout.trim().split('\n');
+      const conf = JSON.parse(lines[lines.length - 1]);
+      t.equal(
+        conf.serviceName,
+        'ns-name',
+        'serviceName was inferred and normalized from package.json',
+      );
+      t.equal(
+        conf.serviceVersion,
+        '1.2.3',
+        'serviceVersion was inferred from package.json',
+      );
+    },
+  },
+  {
+    name: 'pkg-zero-conf-sanitize - serviceName/serviceVersion inferred package.json and sanitized',
+    script: 'fixtures/pkg-zero-conf-sanitize/index.js',
+    cwd: __dirname,
+    noConvenienceConfig: true,
+    checkScriptResult: (t, err, stdout) => {
+      t.error(
+        err,
+        `pkg-zero-conf-sanitize/index.js script succeeded: err=${err}`,
+      );
+      const lines = stdout.trim().split('\n');
+      const conf = JSON.parse(lines[lines.length - 1]);
+      // serviceName sanitization changes any disallowed char to an underscore.
+      // The pkg-zero-conf-sanitize/package.json has a name starting with the
+      // 7 characters that an npm package name can have, but a serviceName
+      // cannot.
+      //    "name": "~*.!'()validNpmName"
+      t.equal(
+        conf.serviceName,
+        '_______validNpmName',
+        'serviceName was inferred and sanitized from package.json',
+      );
+      t.equal(
+        conf.serviceVersion,
+        '1.2.3',
+        'serviceVersion was inferred from package.json',
+      );
+    },
+  },
+  {
+    name: 'pkg-zero-conf-weird - serviceName/serviceVersion inferred package.json and weirdd',
+    script: 'fixtures/pkg-zero-conf-weird/index.js',
+    cwd: __dirname,
+    noConvenienceConfig: true,
+    checkScriptResult: (t, err, stdout) => {
+      t.error(err, `pkg-zero-conf-weird/index.js script succeeded: err=${err}`);
+      const lines = stdout.trim().split('\n');
+      const logs = lines.map((l) => JSON.parse(l));
+      const logWarn = logs.find((log) => log['log.level'] === 'warn');
+      t.ok(
+        logWarn['log.level'] === 'warn' &&
+          logWarn.message.indexOf('serviceName') !== -1,
+        'there is a log.warn about "serviceName"',
+      );
+      const conf = JSON.parse(
+        // Filter out log lines from the APM agent itself. We just want the
+        // `console.log(...)` from the index.js script.
+        lines.filter((ln) => ln.indexOf('"log.level":') === -1)[0],
+      );
+      t.equal(
+        conf.serviceName,
+        'unknown-nodejs-service',
+        'serviceName is the `unknown-{service.agent.name}-service` zero-conf fallback',
+      );
+      t.equal(
+        conf.serviceVersion,
+        '1.2.3',
+        'serviceVersion was inferred from package.json',
+      );
+    },
+  },
+  {
+    name: 'pkg-zero-conf-none - serviceName/serviceVersion resolved to unknown if no package.json present',
+    script: (function () {
+      const path = require('path');
```

Comment:
Note for reviewer: when this [iife](https://developer.mozilla.org/en-US/docs/Glossary/IIFE) executes the requires at the top are not resolved yet. So we need to these ones